Source/WebCore/loader/ContentFilter.cpp

```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
     return !request.isNull();
 }
 
+ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator()
+{
+    ASSERT(m_callback);
+    CheckedPtr contentFilter = m_contentFilter;
+
+    if (m_blocked) {
+        contentFilter->didDecide(State::Blocked);
+        m_callback(ResourceRequest());
+        return;
+    }
+    if (m_platformFiltersAllowed == contentFilter->m_contentFilters.size())
+        contentFilter->didDecide(State::Allowed);
+    m_callback(WTFMove(m_request));
+}
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+{
+    CheckedPtr contentFilter = m_contentFilter;
+
+    if (state == PlatformContentFilter::State::Allowed)
+        ++m_platformFiltersAllowed;
+
+    bool blocked = state == PlatformContentFilter::State::Blocked;
+    if (blocked && !m_blocked) {
+        m_blocked = true;
+        contentFilter->m_blockingContentFilter = platformContentFilter;
+        return;
+    }
+
+    m_request = WTFMove(request);
+}
+
+ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator(ContentFilter& contentFilter, CompletionHandler<void(ResourceRequest&&)>&& callback)
+    : m_contentFilter(contentFilter)
+    , m_callback(WTFMove(callback))
+{
+}
+
+void ContentFilter::continueAfterWillSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
+        continueAfterWillSendRequest(request, redirectResponse);
+        completionHandler(WTFMove(request));
+        return;
```

Comment:
I was originally confused as to why the synchronous version of `continueAfterWillSendRequest` remained, and I think it's just for this.

I think we should rename it to something like `continueAfterWillSendRequestInternal` and move it to private, so outsides (clients and specific platform filters) can't "accidentally" use it anymore.